solr/core/src/java/org/apache/solr/servlet/PriorityBasedRateLimiter.java

```diff
@@ -0,0 +1,127 @@
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import javax.servlet.http.HttpServletRequest;
+import org.apache.solr.common.SolrException;
+import org.apache.solr.core.RateLimiterConfig;
+
+/**
+ * PriorityBasedRateLimiter allocates the slot based on their request priority Currently, it has two
+ * priorities FOREGROUND and BACKGROUND Requests. Client can pass the {@link
+ * org.apache.solr.common.params.CommonParams} SOLR_REQUEST_TYPE_PARAM request header to indicate
+ * the foreground and background request. Foreground requests has high priority than background
+ * requests
+ */
+public class PriorityBasedRateLimiter extends RequestRateLimiter {
+  public static final String SOLR_REQUEST_PRIORITY_PARAM = "Solr-Request-Priority";
+  private final AtomicInteger activeRequests = new AtomicInteger();
+  private final Semaphore numRequestsAllowed;
+
+  private final int totalAllowedRequests;
+
+  private final LinkedBlockingQueue<CountDownLatch> waitingList = new LinkedBlockingQueue<>();
+
+  private final long waitTimeoutInMillis;
+
+  public PriorityBasedRateLimiter(RateLimiterConfig rateLimiterConfig) {
+    super(rateLimiterConfig);
+    this.numRequestsAllowed = new Semaphore(rateLimiterConfig.allowedRequests, true);
+    this.totalAllowedRequests = rateLimiterConfig.allowedRequests;
+    this.waitTimeoutInMillis = rateLimiterConfig.waitForSlotAcquisition;
+  }
+
+  @Override
+  public SlotReservation handleRequest(HttpServletRequest request) {
+    if (!rateLimiterConfig.isEnabled) {
+      return UNLIMITED;
+    }
+    RequestPriorities requestPriority = getRequestPriority(request);
+    if (requestPriority == null) {
+      throw new SolrException(
+          SolrException.ErrorCode.BAD_REQUEST,
+          "Request priority header is not defined or not set properly");
+    }
+    try {
+      if (!acquire(requestPriority)) {
+        return null;
+      }
+    } catch (InterruptedException ie) {
+      return null;
+    }
+    return () -> PriorityBasedRateLimiter.this.release();
+  }
+
+  private boolean acquire(RequestPriorities priority) throws InterruptedException {
+    if (priority.equals(RequestPriorities.FOREGROUND)) {
+      return nextInQueue();
+    } else if (priority.equals(RequestPriorities.BACKGROUND)) {
+      if (this.activeRequests.get() < this.totalAllowedRequests) {
+        return nextInQueue();
+      } else {
+        CountDownLatch wait = new CountDownLatch(1);
+        this.waitingList.put(wait);
+        return wait.await(this.waitTimeoutInMillis, TimeUnit.MILLISECONDS) && nextInQueue();
+      }
+    }
+    return true;
+  }
+
+  private boolean nextInQueue() throws InterruptedException {
+    boolean acquired =
+        this.numRequestsAllowed.tryAcquire(1, this.waitTimeoutInMillis, TimeUnit.MILLISECONDS);
+    if (!acquired) {
+      return false;
+    }
+    this.activeRequests.addAndGet(1);
```

Comment:
I'm pretty sure upon further reflection that you want to increment `activeRequests` _before_ attempting to acquire a permit from the semaphore. Currently it's very unlikely that `activeRequests` would ever go above `totalAllowedRequests`, because incrementing `activeRequests` is protected by the semaphore. This would allow lower-priority requests to "slip through" as under heavy "FOREGROUND" request load, `activeRequests` would mostly by toggling between `totalAllowedRequests` and `totalAllowedRequests - 1`.

If you make this change then of course you'd need to decrement upon _failing_ to acquire.